src/protocol/sort/multi_bit_permutation.rs

```diff
@@ -0,0 +1,206 @@
+///       a. Calculate accumulated `total_sum` = s + `mult_output`
+/// 4. Compute the final output using sum of products executed in parallel for each record.
+#[allow(dead_code)]
+pub async fn multi_bit_permutation<'a, F: Field, S: SecretSharing<F>, C: Context<F, Share = S>>(
+    ctx: C,
+    input: &[Vec<S>],
+) -> Result<Vec<S>, Error> {
+    let num_multi_bits = input.len();
+    assert!(num_multi_bits > 0);
+    let num_records = input[0].len();
+    let num_possible_bit_values = 2 << (num_multi_bits - 1);
+
+    // Equality bit checker: this checks if each secret shared record is equal to any of numbers between 0 and num_possible_bit_values
+    let mut equality_check_futures = Vec::with_capacity(num_possible_bit_values);
+    for j in 0..num_possible_bit_values {
+        let ctx = ctx.clone();
+        equality_check_futures.push(async move { get_bit_equality_checkers(j, input, ctx).await });
+    }
+
+    let equality_checks = try_join_all(equality_check_futures).await?;
+
+    // Compute accumulated sum
+    let mut sop_inputs_transposed = Vec::with_capacity(num_possible_bit_values * num_records);
+    let mut cumulative_sum = S::ZERO;
+    for equality_check in &equality_checks {
+        for check in equality_check {
+            cumulative_sum += check;
+            sop_inputs_transposed.push(cumulative_sum.clone());
+        }
+    }
+
+    // Take sum of products of output of equality check and accumulated sum
+    let mut permutation_futures = Vec::with_capacity(num_records);
+    for rec in 0..num_records {
+        let ctx = ctx.clone();
+        let mut sop_inputs = Vec::with_capacity(num_possible_bit_values);
+        for idx in 0..num_possible_bit_values {
+            sop_inputs.push((
+                &equality_checks[idx][rec],
+                &sop_inputs_transposed[idx * num_records + rec],
+            ));
+        }
+        permutation_futures.push(async move {
+            ctx.sum_of_products(RecordId::from(rec), sop_inputs.as_slice())
+                .await
+        });
+    }
+    try_join_all(permutation_futures).await
+}
+
+/// For a given `idx` check if each of the record has same value as idx.
+/// Steps
+/// 1. Get bit representation of `idx`
+/// 2. Calculate equality check
+///   i. keep record bit if `idx` bit is 1
+///   ii.toggle record bit if idx bit is 0 (Done by taking `share_of_one` - value)
+/// 3. Multiply equality checks for all bits per record - in clear per record this will be 1 only for 1 index while 0 for all others
+async fn get_bit_equality_checkers<F, C, S>(
+    idx: usize,
+    multi_bit_input: &[Vec<S>],
+    ctx: C,
+) -> Result<Vec<S>, Error>
+where
+    F: Field,
+    C: Context<F, Share = S>,
+    S: SecretSharing<F>,
+{
+    let num_multi_bits = multi_bit_input.len();
+    let num_records = multi_bit_input[0].len();
+
+    let ctx_across_bits = ctx.narrow(&MultiplyAcrossBits);
+
+    let idx_in_bits = idx.view_bits::<Lsb0>()[..num_multi_bits].iter().rev();
```

Comment:
you can use MSB order and that should eliminate the need for `rev`

Reply:
hmm, it is actually pretty much the same thing, so I don't think it is worth doing: `view_bits::<Msb0>.iter().skip(idx.size() - num_multi_bits)`

Reply:
agreed.. they both looked similar and this looked a bit cleaner (since we dont do calculation of how much to skip)